File: src/main/java/io/github/numichi/reactive/logger/reactor/MDCSnapshot.java
package io.github.numichi.reactive.logger.reactor;

import org.slf4j.MDC;
import reactor.util.annotation.NonNull;
import reactor.util.annotation.Nullable;

import java.util.Map;

public final class MDCSnapshot implements AutoCloseable {
    private MDCSnapshot(@Nullable final Map<String, String> context) {
        if (context == null) {
            MDC.clear();
        } else {
            MDC.setContextMap(context);
        }
    }
    
    @NonNull
    public static MDCSnapshot of(@Nullable final Map<String, String> context) {
        return new MDCSnapshot(context);
    }
    
    @NonNull
    public static MDCSnapshot empty() {
        return new MDCSnapshot(null);
    }
    
    public Map<String, String> getCopyOfContextMap() {
        return MDC.getCopyOfContextMap();
    }
    
    @Override
    public void close() {
        MDC.clear();
    }
}

File: reactive-logger/src/main/kotlin/io/github/numichi/reactive/logger/reactor/MDCSnapshot.kt
package io.github.numichi.reactive.logger.reactor

import org.slf4j.MDC

class MDCSnapshot private constructor(context: Map<String, String?>?) : AutoCloseable {
    companion object {
        @JvmStatic
        fun of(context: Map<String, String?>?): MDCSnapshot {
            return MDCSnapshot(context)
        }

        @JvmStatic
        fun empty(): MDCSnapshot {
            return MDCSnapshot(null)
        }
    }

    init {
        if (context == null) {
            close()
        } else {
            MDC.setContextMap(context)
        }
    }

    val copyOfContextMap: Map<String, String?>
        get() = MDC.getCopyOfContextMap()

    override fun close() {
        MDC.clear()
    }
}
